### PR TITLE
Add pipeline changes

### PR DESCRIPTION
**Pull request by thomast1906**
### Change description ###
Splitting Pipelines of Azure-Platform-Terraform, a new Azure DevOps Pipeline will be created called "Azure-Platform-Terraform-Global"

Two more additions to pipeline-templates also; the new additional pipeline will be running on the pool hmcts-agent-pool rather than Microsoft hosted with vmImage.

Also, once merged, an additional status check added to repo azure-platform-terraform for the Azure-Platform-Terraform-Global pipeline.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: azure_pipeline-global.yaml [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+            secretsFilter: 'azure-devops-token'
+            runAsPreJob: false
+        - task: Bash@3
+          displayName: Prevent parallel run
+          env:
+            thisbuild: $(Build.BuildId)
+            pipelinedefinition: $(System.DefinitionId)
+          inputs:
+            targetType: 'inline'
+            script: |
```
Comment (timja) [inline azure_pipeline-global.yaml:48]: this should probably be extracted out to a script? possibly even a library?

File: azure_pipeline-global.yaml [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        steps:
+          - task: TerraformInstaller@0
+            inputs:
+              terraformVersion: ${{ variables.terraformVersion }}
+          - task: PowerShell@2
+            displayName: Run tests
+            inputs:
+              targetType: 'inline'
+              script: |
+                Install-Module -Name Pester -MaximumVersion 4.99.99 -Force -Verbose -Scope CurrentUser
```
Comment (timja) [inline azure_pipeline-global.yaml:98]: should be a script so we aren't duplicating this ideally

File: azure_pipeline-global.yaml [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+          environment: 'sbox'
+          component: 'global'
+          service_connection: 'dcd-cftapps-sbox'
+          storage_account_rg: 'core-infra-sbox-rg'
+          storage_account_name: 'cftappssbox'
+          build: $(Build.BuildNumber)
+          agentPool: ${{ variables.agentPool }}
+          terraformVersion: ${{ variables.terraformVersion }}
+
+      - template: pipeline-templates/terraform-plan-global.yaml
```
Comment (timja) [inline azure_pipeline-global.yaml:119]: would be nice to template this https://github.com/hmcts/rdo-sendgrid/blob/master/azure-pipelines.yml#L35

Comment (naikajah) [reply azure_pipeline-global.yaml:119]: yeah, this would be really good to do.

File: azure_pipeline-global.yaml (hunk 1 above)
Comment (thomast1906) [reply azure_pipeline-global.yaml]: Latest merge has included scriptPath for both builds-checks & UnlockTfState-check

File: azure_pipeline-global.yaml (hunk 2 above)
Comment (thomast1906) [reply azure_pipeline-global.yaml]: I didn't add this to a script as it was only a couple of lines. Can add if you want

File: azure_pipeline-global.yaml (hunk 3 above)
Comment (thomast1906) [reply azure_pipeline-global.yaml]: See latest PR - I have went for a templated approach

File: pipeline-scripts/UnlockTFstateIfLocked.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (timja) [inline pipeline-scripts/UnlockTFstateIfLocked.sh:1]: can we not use capitals in the file name? locks weird,

`unlock-tf-state.sh` should be enough

File: pipeline-scripts/builds-check.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+set -e
+organization="TerraformTraining"
+project="TamOpsTerraform"
```
Comment (timja) [inline pipeline-scripts/builds-check.sh:4]: this doesn't look right

**Review by timja — COMMENTED** [review]
Build hasn't run so I can't evaluate a couple of changes i'm not sure about, can you trigger both builds please?

File: pipeline-scripts/UnlockTFstateIfLocked.sh (hunk 4 above)
Comment (thomast1906) [reply pipeline-scripts/UnlockTFstateIfLocked.sh]: added changed to latest commit

File: pipeline-scripts/builds-check.sh (hunk 5 above)
Comment (thomast1906) [reply pipeline-scripts/builds-check.sh]: added changed to latest commit

File: pipeline-templates/terraform.yaml [hunk 6]
```diff
@@ -18,19 +36,38 @@ jobs:
...
-        artifact: ${{ parameters.environment }}${{ parameters.component }}${{ parameters.build }}plan
+
+      - task: AzureCLI@2
+        displayName: 'Check if TF state locked'
+        condition: always()
+        inputs:
+          scriptType: bash
+          scriptLocation: scriptPath
+          azureSubscription: ${{ parameters.service_connection }}
+          scriptPath: $(System.DefaultWorkingDirectory)/pipeline-scripts/unlock-tf-state.sh
```
Comment (timja) [inline pipeline-templates/terraform.yaml:72]: this file doesn't seem to exist

File: azure_pipeline-global.yaml [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      component: 'global'
+      service_connection: 'dcd-cftapps-prod'
+      storage_account_rg: 'core-infra-prod-rg'
+      storage_account_name: 'cftappsprod'
+      dependsOn: 'sbox_global'
+
+stages:
+  - stage: Precheck
+    jobs:
+      - job: Validate
```
Comment (timja) [inline azure_pipeline-global.yaml:100]: seems to be missing validation steps?

File: pipeline-templates/terraform.yaml [hunk 8]
```diff
@@ -1,11 +1,29 @@
...
+    - job: TerraformPlanAndApply
+      pool: ${{ parameters.pool }}
+      timeoutInMinutes: ${{ parameters.timeoutInMinutes }}
+      steps:     
       - task: TerraformInstaller@0
+        displayName: Terraform install
         inputs:
           terraformVersion: ${{ parameters.terraformVersion }}
+
+      - task: PowerShell@2
```
Comment (timja) [inline pipeline-templates/terraform.yaml:11]: this only needs to be run once per pipeline not on every terraform run.

**Review by timja — CHANGES_REQUESTED** [review]

File: pipeline-templates/terraform.yaml (hunk 6 above)
Comment (thomast1906) [reply pipeline-templates/terraform.yaml:72]: yeah, had renamed as per above and was delayed with the merge for it.

File: pipeline-templates/terraform.yaml (hunk 8 above)
Comment (timja) [reply pipeline-templates/terraform.yaml:11]: this is taking 15-20 seconds on each job right now

Comment (thomast1906) [reply pipeline-templates/terraform.yaml]: Added to the current "validate" stage

File: azure_pipeline-global.yaml (hunk 7 above)
Comment (thomast1906) [reply azure_pipeline-global.yaml:100]: Which steps are missing?

File: azure_pipeline.yaml [hunk 9]
```diff
@@ -6,18 +6,164 @@ trigger:
...
+
+    - deployment: 'sbox_compliance'
+      environment: 'sbox'
+      component: 'compliance'
+      service_connection: 'dcd-cftapps-sbox'
+      storage_account_rg: 'core-infra-sbox-rg'
+      storage_account_name: 'cftappssbox'
+      dependsOn: 'Precheck'
+    
+    # - deployment: 'ptl_dynatrace_activegate'
```
Comment (timja) [inline azure_pipeline.yaml:57]: why all this is commented out?

File: azure_pipeline-global.yaml [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            pipelinedefinition: $(System.DefinitionId)
+            azuredevopstoken: $(azure-devops-token)
+          inputs:
+            filePath: pipeline-scripts/builds-check.sh
+        - task: PowerShell@2
+          displayName: Run tests
+          inputs:
+            targetType: 'inline'
+            script: |
+              Install-Module -Name Pester -MaximumVersion 4.99.99 -Force -Verbose -Scope CurrentUser
```
Comment (timja) [inline azure_pipeline-global.yaml:123]: could we  not duplicate this please?, we need to upgrade pester at some point and we shouldn't have to update multiple places in a repo

Comment (thomast1906) [reply azure_pipeline-global.yaml]: Swapped to filePath

File: azure_pipeline.yaml (hunk 9 above)
Comment (thomast1906) [reply azure_pipeline.yaml]: Was running a quick test on some envs - all uncommented now.

File: azure_pipeline-global.yaml (hunk 7 above)
Comment (timja) [reply azure_pipeline-global.yaml:100]: terraform fmt checks

File: azure_pipeline.yaml [hunk 11]
```diff
@@ -6,18 +6,164 @@ trigger:
...
+
 variables:
   terraformVersion: 0.12.26
   agentPool: 'ubuntu-18.04'
+  build: $(Build.BuildNumber)
+  isMaster: $[eq(variables['Build.SourceBranch'], 'refs/heads/master')]
   action:
 
+parameters:
+  - name: environments
```
Comment (timja) [inline azure_pipeline.yaml:22]: is environments the right term for this object?

File: azure_pipeline.yaml [hunk 12]
```diff
@@ -6,18 +6,164 @@ trigger:
...
+      storage_account_name: 'cftappsprod'
+      dependsOn: 'sbox_cftapps_cluster_lb'
+
 stages:
   - stage: Precheck
     jobs:
       - job: Validate
-        pool:
-          vmImage: $(agentPool)
+        pool:
```
Comment (timja) [inline azure_pipeline.yaml:165]:
```suggestion
        pool:
```

**Review by timja — APPROVED** [review]
main issue left is environment name in list, as it's not actually an environment but nothing blocking on my side anymore

File: azure_pipeline.yaml (hunk 12 above)
Comment (thomast1906) [reply azure_pipeline.yaml:165]: vmImage is required if we want to keep as it is with the reference to Microsoft hosted image:- ubuntu-18.04

If just using pool, it fails, 
"The pipeline is not valid. Could not find a pool with name ubuntu-18.04"

File: azure_pipeline.yaml (hunk 11 above)
Comment (thomast1906) [reply azure_pipeline.yaml]: changed "environments" to "environment_components"

File: pipeline-templates/terraform.yaml [hunk 13]
```diff
@@ -18,19 +20,38 @@ jobs:
...
         inputs:
           targetType: 'inline'
           workingDirectory: '$(System.DefaultWorkingDirectory)/components/${{ parameters.component }}'
           script: |
             rm -f terraform.tfstate* && rm -rf .terraform/
-      - publish: $(System.DefaultWorkingDirectory)/components/${{ parameters.component }}/${{ parameters.environment }}${{ parameters.component }}${{ parameters.build }}plan
-        artifact: ${{ parameters.environment }}${{ parameters.component }}${{ parameters.build }}plan
+
+      - task: AzureCLI@2
+        displayName: 'Check if TF state locked'
```
Comment (timja) [inline pipeline-templates/terraform.yaml:50]: maybe? the script is doing more than checking
```suggestion
        displayName: 'Unlock TF state if required'
```

**Review by timja — APPROVED** [review]

**Pull request merged**